analyses/npl/05_station_correlation.md

```diff
@@ -0,0 +1,365 @@
+```python
+from pathlib import Path
+import os
+import sys
+from collections import Counter
+
+import pandas as pd
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+from matplotlib.ticker import MaxNLocator
+import numpy as np
+from scipy.signal import correlate
+from scipy.interpolate import interp1d
+
+path_mod = f"{Path(os.path.dirname(os.path.realpath(''))).parents[0]}/"
+sys.path.append(path_mod)
+
+from src.indicators.flooding.glofas import utils, glofas
+```
+
+```python
+COUNTRY_ISO3 = 'npl'
+
+DURATION = 1 # Number of days that the GloFAS threshold should be exceeded
+
+RP_LIST = [1.5, 2, 3, 5, 10, 20] # List of return period values to compare
+
+STATIONS = {
+    'Koshi': ['Chatara', 'Simle', 'Majhitar', 'Kampughat'],
+    'Karnali': ['Chisapani', 'Asaraghat', 'Dipayal', 'Samajhighat'],
+    'Rapti': ['Kusum'],
+    'Bagmati': ['Rai_goan'],
+    'Babai': ['Chepang']
+}
+STATIONS_BY_MAJOR_BASIN = {
+    'Koshi': ['Chatara', 'Simle', 'Majhitar', 'Kampughat', 'Rai_goan'],
+    'Karnali': ['Chisapani', 'Asaraghat', 'Dipayal', 'Samajhighat', 'Kusum', 'Chepang'],
+}
+```
+
+```python
+ds_glofas_reanalysis = utils.get_glofas_reanalysis(
+    country_iso3=COUNTRY_ISO3)
+df_return_period = utils.get_return_periods(ds_glofas_reanalysis, RP_LIST)
+```
+
+## When do activations occur at the different stations
+
+We want to know how much stations within each basin are correlated, and if activation events in the past have occured simultaneously, so that we could have a multi-station trigger
+
+```python
+cmap = mpl.cm.get_cmap('plasma_r')
+clist = cmap(np.linspace(0, 1, len(RP_LIST)))
+cdict = {rp: c for rp, c in zip(RP_LIST, clist)}
+legend_title = 'RP'
+```
+
+```python
+events_all = {}
+
+for basin in ['Koshi', 'Karnali', 'Rapti', 'Bagmati', 'Babai']:
+    
+    station_events = {}   
+    stations = STATIONS[basin]
+    
+    for istation, station in enumerate(stations):  
+        
+        rp_events = {}
+        
+        for rp in RP_LIST:
+            rp_val=df_return_period.loc[rp, station]
+            df_activations = utils.get_glofas_activations(ds_glofas_reanalysis[station], rp_val, DURATION)
+            rp_events[rp] = df_activations
+            
+        station_events[station] = rp_events
+        
+    events_all[basin] = station_events
+```
+
+```python
+x = np.array([1, 2, 3, 4, 4, 3, 2, 3, 4, 3, 2])
+g = utils.get_groups_above_threshold(x, 3.5)
+
+```
+
+### Plot river discharge vs time for all stations
```

Comment:
I really like these plots! This might be good to add to utils as a generalized plotting function when we have more time...

Reply:
Glad you like them! Yes I can definitely generalize this, especially if I find myself using it for another pilot